collector/filefd_linux.go

```diff
@@ -0,0 +1,94 @@
+// +build !nonetstat
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+)
+
+const (
+	procFileFDStat      = "/proc/sys/fs/file-nr"
+	fileFDStatSubsystem = "filefd"
+)
+
+type fileFDStatCollector struct {
+	metrics map[string]prometheus.Gauge
+}
+
+func init() {
+	Factories[fileFDStatSubsystem] = NewFileFDStatCollector
+}
+
+// NewFileFDStatCollector returns a new Collector exposing file-nr stats
+func NewFileFDStatCollector() (Collector, error) {
+	return &fileFDStatCollector{
+		metrics: map[string]prometheus.Gauge{},
+	}, nil
+}
+
+func (c *fileFDStatCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	fileNRStats, err := getFileFDStats(procFileFDStat)
+	if err != nil {
+		return fmt.Errorf("couldn't get file-nr: %s", err)
+	}
+	for protocol, protocolStats := range fileNRStats {
```

Comment:
For this collector that just sets three well-known metrics, it would make more sense to simply manually define the three metrics upfront and set them to their right values on `Update()`.